server/msteams/client.go

```diff
@@ -0,0 +1,416 @@
+		clientType:   "bot",
+		tenantId:     tenantId,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+		botUsername:  botUsername,
+		botPassword:  botPassword,
+	}
+}
+
+func (tc *ClientImpl) Connect() error {
+	var ts oauth2.TokenSource
+	if tc.clientType == "bot" {
+		var err error
+		m := msauth.NewManager()
+		ts, err = m.ResourceOwnerPasswordGrant(
+			tc.ctx,
+			tc.tenantId,
+			tc.clientId,
+			tc.clientSecret,
+			tc.botUsername,
+			tc.botPassword,
+			[]string{teamsDefaultScope},
+		)
+		if err != nil {
+			return err
+		}
+	} else if tc.clientType == "app" {
+		var err error
+		m := msauth.NewManager()
+		ts, err = m.ClientCredentialsGrant(
+			tc.ctx,
+			tc.tenantId,
+			tc.clientId,
+			tc.clientSecret,
+			[]string{teamsDefaultScope},
+		)
+		if err != nil {
+			return err
+		}
+	} else {
+		panic("Not valid client type, this shouldn't happen ever.")
+	}
+
+	httpClient := oauth2.NewClient(tc.ctx, ts)
+	graphClient := msgraph.NewClient(httpClient)
+	tc.client = graphClient
+
+	if tc.clientType == "bot" {
+		req := graphClient.Me().Request()
+		r, err := req.Get(tc.ctx)
+		if err != nil {
+			return err
+		}
+		tc.botID = *r.ID
+	}
+
+	return nil
+}
+
+func (tc *ClientImpl) SendMessage(teamID, channelID, parentID, message string) (string, error) {
+	content := &msgraph.ItemBody{Content: &message}
+	rmsg := &msgraph.ChatMessage{Body: content}
+
+	var res *msgraph.ChatMessage
+	if len(parentID) > 0 {
+		var err error
+		ct := tc.client.Teams().ID(teamID).Channels().ID(channelID).Messages().ID(parentID).Replies().Request()
+		res, err = ct.Add(tc.ctx, rmsg)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		var err error
+		ct := tc.client.Teams().ID(teamID).Channels().ID(channelID).Messages().Request()
+		res, err = ct.Add(tc.ctx, rmsg)
+		if err != nil {
+			return "", err
+		}
+	}
+	return *res.ID, nil
+}
+
+func (tc *ClientImpl) SubscribeToChannel(teamID, channelID, notificationURL string, webhookSecret string) (string, error) {
+	resource := "teams/" + teamID + "/channels/" + channelID + "/messages"
+	expirationDateTime := time.Now().Add(60 * time.Minute)
+	changeType := "created"
+	subscription := msgraph.Subscription{
+		Resource:           &resource,
+		ExpirationDateTime: &expirationDateTime,
+		NotificationURL:    &notificationURL,
+		ClientState:        &webhookSecret,
+		ChangeType:         &changeType,
+	}
+	ct := tc.client.Subscriptions().Request()
+	res, err := ct.Add(tc.ctx, &subscription)
+	if err != nil {
+		return "", err
+	}
+	return *res.ID, nil
+}
+
+func (tc *ClientImpl) RefreshSubscriptionPeriodically(ctx context.Context, subscriptionID string) error {
+	for {
+		select {
+		case <-time.After(time.Minute):
+			expirationDateTime := time.Now().Add(10 * time.Minute)
+			updatedSubscription := msgraph.Subscription{
+				ExpirationDateTime: &expirationDateTime,
+			}
+			ct := tc.client.Subscriptions().ID(subscriptionID).Request()
+			err := ct.Update(tc.ctx, &updatedSubscription)
```

Comment:
Just so I understand, is this something recommended by Microsoft's docs?

Reply:
Yes, exactly. I need to refresh it from time to time the subscription to keep it alive. If not, Microsoft teams is going to stop sending me the data.